src/trg-client.c

```diff
@@ -271,6 +277,137 @@ static GList *trg_client_headers_array_to_list(JsonArray *array)
+    data->host = g_strdup(host);
+    data->results = g_hash_table_new_full(g_str_hash, g_str_equal,
+                                          g_free, NULL);
+    data->timer = g_timer_new();
+    return data;
+}
+
+static void
+mdns_lookup_data_free(mDNSLookupData *data)
+{
+    if (!data)
+        return;
+    g_clear_pointer(&data->host, g_free);
+    g_clear_pointer(&data->timer, g_timer_destroy);
+    g_clear_pointer(&data->results, g_hash_table_destroy);
+    g_free (data);
+}
+
+static bool
+add_results(GString *result, mDNSLookupData *data, guint num_results)
+{
+    GList *keys, *l;
+
+    keys = g_hash_table_get_keys(data->results);
+    if (!keys)
+        return num_results;
+    for (l = keys; l != NULL; l = l->next) {
+        const char *ip = l->data;
+
+        if (num_results > 0)
+            g_string_append_c(result, ',');
+        g_string_append(result, ip);
+        num_results++;
+    }
+    g_list_free(keys);
+    return num_results;
+}
+
+static bool
+lookup_stop_cb(void *cookie)
+{
+    mDNSLookupData *data = cookie;
+
+    return (g_hash_table_size(data->results) > 0 ||
+            g_timer_elapsed(data->timer, NULL) > (double) LOOKUP_TIMEOUT);
+}
+
+static void
+lookup_listen_cb(void *cookie, int status, const struct rr_entry *entries)
+{
+    struct rr_entry *entry;
+    mDNSLookupData *data = cookie;
+
+    if (status < 0)
+            return;
+    entry = (struct rr_entry *) entries;
+    while (entry) {
+        if (entry->type == RR_A)
+            g_hash_table_insert(data->results, g_strdup(entry->data.A.addr_str), GINT_TO_POINTER(1));
+        else if (entry->type == RR_AAAA)
+            g_hash_table_insert(data->results, g_strdup(entry->data.AAAA.addr_str), GINT_TO_POINTER(1));
+        entry = entry->next;
+    }
+}
+
+static char *
+get_hosts_for_local(const char *host, int port)
+{
+    struct mdns_ctx *ctx;
+    const char *names[] = { NULL };
+    mDNSLookupData *data = NULL;
+    GString *result = NULL;
+    guint num_results = 0;
+    int r;
+
+    if (!g_str_has_suffix(host, ".local") &&
+        !g_str_has_suffix(host, ".local."))
+        return NULL;
+
+    if ((r = mdns_init(&ctx, NULL, MDNS_PORT)) < 0)
+        goto err;
+
+    result = g_string_new(NULL);
+    g_string_append_printf(result, "+%s:%d:", host, port);
+
+    names[0] = host;
+    /* IPv4 */
+    data = mdns_lookup_data_new(host);
+    if ((r = mdns_listen(ctx, names, G_N_ELEMENTS(names), RR_A, LOOKUP_TIMEOUT, lookup_stop_cb,
+                     lookup_listen_cb, data)) < 0)
+        goto err;
+    num_results = add_results(result, data, num_results);
+    mdns_lookup_data_free(data);
+
+    /* IPv6 */
+    data = mdns_lookup_data_new(host);
+    if ((r = mdns_listen(ctx, names, G_N_ELEMENTS(names), RR_AAAA, LOOKUP_TIMEOUT, lookup_stop_cb,
+                     lookup_listen_cb, data)) < 0)
+        goto err;
+    num_results = add_results(result, data, num_results);
+    if (num_results == 0)
```

Comment:
I find this unclear. This seems like not an error. Maybe rename `err` to `cleanup` or something more generic.